### ISSUE TITLE
Android hot reload cause crash

### ISSUE THREAD

**Issue by bsutton** [Canardoux/tau#304]
The TrackPlayerPlugin and the FlautoPlayerPlugin share a common slot array.

If you do a hot reload the array is not cleared out.
The result is that when you restart the slots are not reset and you can get a class cast exception (as a given slot could contain a Track or a Fluato) plugin.

Whilst this is not a problem for running code it will cause problems and confusing during development.
We have a couple of options.
1) separate the slot arrays so each plugin has its own.
2) clean up the slot arrays when a hot reload occurs.
- I have some code in Playbar that can detect a hot reload however there is no code in the plugins to free slots (the freeSlot method exists but is never called).
I thought that onDetachedFromEngine from engine in Flauto.java might be the place to do this but I have no experience with plugins.

To reproduce this problem:
Play Asset with OS Media Player off
hot reload
Play Asset with OS Media Player on.

This problem can be reproduced on the dev branch code however I believe that it will be a problem in the current production release.

```

I/flutter (23913): Loading starting
I/flutter (23913): Calling startPlayer
I/flutter (23913): regisetered Proxy: 1065489832
I/flutter (23913): initialise called Proxy: 1065489832
I/flutter (23913): #################SpinKit rendered###################
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): Failed to handle method call
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): java.lang.ClassCastException: com.dooboolab.fluttersound.FlutterSoundPlayer cannot be cast to com.dooboolab.fluttersound.TrackPlayer
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at com.dooboolab.fluttersound.TrackPlayerPlugin.onMethodCall(TrackPlayer.java:132)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:231)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:93)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:642)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at android.os.Looper.loop(Looper.java:174)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/MethodChannel#com.dooboolab.flutter_sound_track_player(23913): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

**Comment by Larpoux** [comment]
My plan was to use the same slot array, even with `RecordPlayer`.
Classes `RecordPlayer`, `TrackPlayer`, and `FlutterSoundPlayer` would inherit from a common base class which will do the **mux** on its flutter channel. No duplicated code anymore.

But if this is a problem with hot reload we can reconsider that.

**Comment by bsutton** [comment]
I think there is perhaps an argument for a single plugin which then routes
calls to the appropriate class.

At the very least I think we should combine the trackplayer and soundplayer
into a single plug.
I can't see any obvious reason why these are separate (unless there is some
aspect I've missed).
Tracker already extends soundplayer.  A singular class would make the
plugin easier to work with.

**Comment by Larpoux** [comment]
I have not looked to this Hot-Reload very seriously, but my thinking is that your actual proposition on the `dev` branch is correct : 

Have the three "plugins" inherited from a base class which manage the sessions slots.
For me this is clean code.

**Comment by tuanbs** [comment]
This issue still occurs on the latest version `4.0.1+1`.

**Comment by Larpoux** [comment]
See #387

**Comment by Larpoux** [comment]
Fixed in Flutter Sound release 7.6.3
